lib/services/watcher.go

```diff
@@ -1666,3 +1666,159 @@ func (p *accessRequestCollector) processEventAndUpdateCurrent(ctx context.Contex
+}
+
+// SAMLIdPServiceProviderWatcher is built on top of resourceWatcher to monitor SAML IdP service provider resources.
+type SAMLIdPServiceProviderWatcher struct {
+	*resourceWatcher
+	*samlIDPServiceProviderCollector
+}
+
+// samlIDPServiceProviderCollector accompanies resourceWatcher when monitoring SAML IdP service provider resources.
+type samlIDPServiceProviderCollector struct {
+	// SAMLIdPServiceProviderWatcherConfig is the watcher configuration.
+	SAMLIdPServiceProviderWatcherConfig
+	// current holds a map of the currently known SAML IdP service provider resources.
+	current map[string]types.SAMLIdPServiceProvider
+	// lock protects the "current" map.
+	lock sync.RWMutex
+	// initializationC is used to check that the watcher has been initialized properly.
+	initializationC chan struct{}
+	once            sync.Once
+}
+
+// resourceKind specifies the resource kind to watch.
+func (p *samlIDPServiceProviderCollector) resourceKind() string {
+	return types.KindSAMLIdPServiceProvider
+}
+
+// initializationChan is used to check if the initial state sync has been completed.
+func (p *samlIDPServiceProviderCollector) initializationChan() <-chan struct{} {
+	return p.initializationC
+}
+
+// getResourcesAndUpdateCurrent refreshes the list of current resources.
+func (p *samlIDPServiceProviderCollector) getResourcesAndUpdateCurrent(ctx context.Context) error {
+	var samlIDPServiceProviders []types.SAMLIdPServiceProvider
+	nextToken := ""
+	for {
+		var sps []types.SAMLIdPServiceProvider
+		var err error
+		sps, nextToken, err = p.SAMLIdPServiceProviders.ListSAMLIdPServiceProviders(ctx, p.PageSize, nextToken)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		samlIDPServiceProviders = append(samlIDPServiceProviders, sps...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	newCurrent := make(map[string]types.SAMLIdPServiceProvider, len(samlIDPServiceProviders))
+	for _, samlIDPServiceProvider := range samlIDPServiceProviders {
+		newCurrent[samlIDPServiceProvider.GetName()] = samlIDPServiceProvider
+	}
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	p.current = newCurrent
+	p.defineCollectorAsInitialized()
+
+	select {
+	case <-ctx.Done():
+		return trace.Wrap(ctx.Err())
+	case p.SAMLIdPServiceProvidersC <- samlIDPServiceProviders:
+	}
+
+	return nil
+}
+
+func (p *samlIDPServiceProviderCollector) defineCollectorAsInitialized() {
+	p.once.Do(func() {
+		// mark watcher as initialized.
+		close(p.initializationC)
+	})
+}
+
+// processEventAndUpdateCurrent is called when a watcher event is received.
+func (p *samlIDPServiceProviderCollector) processEventAndUpdateCurrent(ctx context.Context, event types.Event) {
+	if event.Resource == nil || event.Resource.GetKind() != types.KindSAMLIdPServiceProvider {
+		p.Log.Warnf("Unexpected event: %v.", event)
+		return
+	}
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	switch event.Type {
+	case types.OpDelete:
+		delete(p.current, event.Resource.GetName())
+		select {
```

Comment:
Ditto for holding the lock for the duration of the select, here and in the next case.

Reply:
I disagree, it still looks like we might hold the lock for too long. Could we grab it solely during the `p.current` changes, on the spot, and release asap?